src/ctia/entity/vulnerability/cpe.clj

```diff
@@ -0,0 +1,218 @@
+
+(defn ->internal
+  "Replaces all escaped characters in String `s` with internal representations"
+  [s]
+  (reduce-kv (fn ->internal-representation-rf [acc escape internal-representation]
+               (str/replace acc escape internal-representation))
+             s
+             escape->internal))
+
+(defn internal->triple-escape
+  "Replaces all internal representations in String `s` with their
+  triple escaped forms."
+  [s]
+  (reduce-kv (fn internal->triple-escape-rf [acc escape internal-representation]
+               (str/replace acc internal-representation (str "\\\\" escape)))
+             s
+             escape->internal))
+
+(defn str->pattern
+  "Converts String `s` with CPE Match String wildcards to a
+  `java.util.regex.Pattern`."
+  [s]
+  (-> s
+      (str/replace "*" ".*")
+      (str/replace "?" ".?")
+      re-pattern))
+
+(defn component-match?
+  "Returns true when CPE Match String components `s1` and `s2` match,
+  otherwise false. When `s1` or `s2` is set to `*` or `-` returns
+  true, indicating a match of `ANY` and `Not Applicable`
+  respectively."
+  [s1 s2]
+  (let [p1 (str->pattern s1)
+        p2 (str->pattern s2)]
+    (cond
+      (= s1 s2) true
+      (or (and (= s1 "-") (= s2 "*"))
+          (and (= s1 "*") (= s2 "-"))) true
+      (or (re-matches p1 s2) (re-matches p2 s1)) true
+      :else false)))
+
+(defn satisfies-interval?
+  "Returns true when version satisfies interval, otherwise
+  false. Returns true when version is set to `*` or `-`, which
+  indicate `ANY` and `Not Applicable` respectively."
+  [version
+   {:keys [version-start-including
+           version-start-excluding
+           version-end-including
+           version-end-excluding]}]
+  (letfn [(closed [] (and (left-closed) (right-closed)))
+          (open [] (and (left-open) (right-open)))
+          (left-closed-right-open [] (and (left-closed) (right-open)))
+          (left-open-right-closed [] (and (left-open) (right-closed)))
+          (left-closed [] (v/older-or-equal? version-start-including version))
+          (left-open [] (v/older? version-start-excluding version))
+          (right-closed [] (v/newer-or-equal? version-end-including version))
+          (right-open [] (v/newer? version-end-excluding version))]
+    (cond
+      (or (= version "*") (= version "-")) true
+      (and version-start-including version-end-including) (closed)
+      (and version-start-excluding version-end-excluding) (open)
+      (and version-start-including version-end-excluding) (left-closed-right-open)
+      (and version-start-excluding version-end-including) (left-open-right-closed)
+      version-start-including (left-closed)
+      version-start-excluding (left-open)
+      version-end-including (right-closed)
+      version-end-excluding (right-open)
+      :else true)))
+
+(defprotocol ICpeMatch
+  (match? [this CpeMatch])
+  (required-components-specified? [this]))
+
+(defrecord CpeMatch [part vendor product version version_update
```

Comment:
In such a case, in IROH services we tend to limit the use of `defprotocol`. Here is an extract from the [Contributing.org](https://github.com/threatgrid/ctia/blob/master/CONTRIBUTING.org#guidelines):
`Only use defprotocol for declaring TK services. Never use it unless you can’t do otherwise. Most of the time using case and/or cond will be preferable.`
After a better review, I have the feeling that the use of this `defrecord` could be replaced by a schema with the required constraints and the `match?` function:


``` 
(defn components-specified? [component]
    (and (not= component "*") (not= component "-")))
    
(s/defschema CpeMatch
   {:part (s/enum "a" "h" "o")
    :vendor (s/pred components-specified?)
    :product (s/pred components-specified?)
    :version s/Str
    :version_update s/Str
    ...
    })
    
    (s/defn match? :- s/boolean
     [{:keys [version ...]} :- CPEMatch
      cpe-match :- CpeMatch]
    (and (satisfies-interval? version cpe-match)
         (and cpe-match
              (component-match? part (:part cpe-match))
              (component-match? vendor (:vendor cpe-match))
              (component-match? product (:product cpe-match))
              (component-match? version (:version cpe-match))
              (component-match? version_update (:version_update cpe-match))
              (component-match? edition (:edition cpe-match))
              (component-match? lang (:lang cpe-match))
              (component-match? sw_edition (:sw_edition cpe-match))
              (component-match? target_sw (:target_sw cpe-match))
              (component-match? target_hw (:target_hw cpe-match))
              (component-match? other (:other cpe-match)))))
```